namui/rust-ffmpeg/src/format/mod.rs

```diff
@@ -167,6 +178,83 @@ pub fn input<P: AsRef<Path>>(path: &P) -> Result<context::Input, Error> {
     }
 }
 
+pub fn input_bytes(bytes: Arc<Vec<u8>>) -> Result<context::Input, Error> {
+    unsafe {
+        unsafe extern "C" fn read_packet(
+            opaque: *mut c_void,
+            buffer: *mut u8,
+            _buffer_size: i32,
+        ) -> i32 {
+            let cursor = &mut *(opaque as *mut Cursor<InputBytes>);
+            let buffer = &mut *(buffer as *mut [u8; INPUT_BUFFER_SIZE]);
+            match cursor.read(buffer) {
+                Ok(read) => read as i32,
+                Err(_) => AVERROR_INVALIDDATA,
+            }
+        }
+        unsafe extern "C" fn seek(opaque: *mut c_void, offset: i64, whence: i32) -> i64 {
+            let cursor = &mut *(opaque as *mut Cursor<InputBytes>);
+            let seek_from = match whence {
+                SEEK_SET => SeekFrom::Start(offset as u64),
+                SEEK_CUR => SeekFrom::Current(offset),
+                SEEK_END => SeekFrom::End(offset),
+                AVSEEK_SIZE => return cursor.get_ref().0.len() as i64,
+                _ => return -1,
+            };
+            match cursor.seek(seek_from) {
+                Ok(position) => position as i64,
+                Err(_error) => -1,
+            }
+        }
+        let buffer: *mut c_void = av_malloc(INPUT_BUFFER_SIZE);
+        let bytes: *mut Cursor<InputBytes> =
+            Box::into_raw(Box::new(Cursor::new(InputBytes(bytes))));
+        let file_name = CString::new("").unwrap();
+
+        let mut ps = avformat_alloc_context();
+        let avio = avio_alloc_context(
+            buffer as *mut u8,
+            INPUT_BUFFER_SIZE as i32,
+            0,
+            bytes as *mut _,
+            Some(read_packet),
+            None,
+            Some(seek),
+        );
+
+        (*ps).pb = avio;
+        (*ps).flags = AVFMT_FLAG_CUSTOM_IO;
+
+        let free = || {
+            av_free(buffer);
+            bytes.drop_in_place();
+        };
+
+        match avformat_open_input(
+            &mut ps,
+            file_name.as_ptr(),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        ) {
+            0 => match avformat_find_stream_info(ps, ptr::null_mut()) {
+                r if r >= 0 => Ok(context::Input::wrap(
+                    ps,
+                    context::input::Mode::Bytes { input_bytes: bytes },
+                )),
+                e => {
+                    avformat_close_input(&mut ps);
+                    free();
+                    Err(Error::from(e))
+                }
+            },
+            e => {
+                free();
+                Err(Error::from(e))
+            }
+        }
+    }
+}
+
```

Comment:
I think it's not a good idea to update rust-ffmpeg directly because then we cannot pull their code for update.
Can you impl this in namui-media? I guess you can do with rust-ffmpeg and rust-ffmpeg-sys.

Reply:
Let me check

Reply:
All ffmpeg changes were reverted.

Reply:
Ready to review?

Reply:
yes